homeassistant/components/hunterdouglas_powerview/__init__.py

```diff
@@ -97,21 +98,41 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
     async def async_update_data():
         """Fetch data from shade endpoint."""
-        async with async_timeout.timeout(10):
-            shade_entries = await shades.get_resources()
-        if not shade_entries:
-            raise UpdateFailed("Failed to fetch new shade data.")
-        return _async_map_data_by_id(shade_entries[SHADE_DATA])
+
+        if coordinator.update_interval == UPDATE_INTERVAL_MAINTENANCE:
+            _LOGGER.debug("Polling returned to %s", UPDATE_INTERVAL_DEFAULT)
+            coordinator.update_interval = UPDATE_INTERVAL_DEFAULT
+
+        try:
+            async with async_timeout.timeout(10):
+                shade_entries = await shades.get_resources()
+            if not shade_entries or isinstance(shade_entries, bool):
+                # hub returns boolean on a 204/423 empty response (maintenance)
+                # continual polling results in inevitable error
+                # restart of hub takes between 3-5 minutes and generally between 12am-3am
+                _LOGGER.debug(
+                    "Hub is reporting that maintenance is underway. Pausing polling for %s",
+                    UPDATE_INTERVAL_MAINTENANCE,
+                )
+                coordinator.update_interval = UPDATE_INTERVAL_MAINTENANCE
+                return
+
+            # moved inside try to prevent attempting to access empty index on error
+            return _async_map_data_by_id(shade_entries[SHADE_DATA])
```

Comment:
Please move code that isn't expected to raise `HUB_EXCEPTIONS` outside of the try

Reply:
i think all this code needs to be inside the TRY because it can throw an exception
or are you suggesting moving 109-118 outside the try ?

if i moved that outside the try then line 121 would return (and throw an error) where i need the above section to identify that the hub is undergoing maintenance

Reply:
I think you can write it something like this.

```python3
    async def async_update_data():
        """Fetch data from shade endpoint."""
        if coordinator.update_interval == UPDATE_INTERVAL_MAINTENANCE:
            _LOGGER.debug("Polling returned to %s", UPDATE_INTERVAL_DEFAULT)
            coordinator.update_interval = UPDATE_INTERVAL_DEFAULT

        try:
            async with async_timeout.timeout(10):
                shade_entries = await shades.get_resources()
        except HUB_EXCEPTIONS as err:
            raise UpdateFailed(f"Failed to fetch new shade data. {err}") from err

        if not shade_entries or isinstance(shade_entries, bool):
            # hub returns boolean on a 204/423 empty response (maintenance)
            # continual polling results in inevitable error
            # restart of hub takes between 3-5 minutes and generally between 12am-3am
            _LOGGER.debug(
                "Hub is reporting that maintenance is underway. Pausing polling for %s",
                UPDATE_INTERVAL_MAINTENANCE,
            )
            coordinator.update_interval = UPDATE_INTERVAL_MAINTENANCE
            # Return the cached data while maintenance is happening
            return coordinator.data

        try:
            # moved inside try to prevent attempting to access empty index on error
            data = _async_map_data_by_id(shade_entries[SHADE_DATA])
        except IndexError as ex:
            raise UpdateFailed(f"Shade data missing from hub: {ex}") as ex

        return data
```